### PR TITLE
Feature/doctor-preview-test

### PR DESCRIPTION
**Pull request by AntonMishyn**
develop

## GitHub Board

**Issue link**

[#34 BE - doctors preview task (main page) ](https://github.com/ita-social-projects/dokazovi-be/issues/34)

**Story link**

[#13 Story](https://github.com/ita-social-projects/dokazovi-be/issues/13)

## Code reviewers
- [ ] @melnikovae87

## Summary of issue

1. - [x] DTO for doctor (and latest post from him)
2. - [x] API (POST) controller with request configuration
3. - [x] Service layer
4. - [x] Repository layer (sort by making posts)
5. - [x] Unit tests

## Summary of change

- [x]  Refactor Entities
- [x]  Implement UnitTests for UserController, mappers, UserService.

### CODE REVIEW

File: src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java [hunk 1]
```diff
@@ -36,25 +34,26 @@
 	@OneToMany(mappedBy = "mainDirection")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<PostEntity> mainPostsDirections;
+	private Set<PostEntity> mainPostsDirections = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java:37]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java [hunk 2]
```diff
@@ -36,25 +34,26 @@
...
 
 	@ManyToMany(mappedBy = "directions")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<PostEntity> posts;
+	private Set<PostEntity> posts = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java:42]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java [hunk 3]
```diff
@@ -36,25 +34,26 @@
...
 
 	@OneToMany(mappedBy = "mainDirection")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<UserEntity> mainUsersDirections;
+	private Set<UserEntity> mainUsersDirections = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java:47]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java [hunk 4]
```diff
@@ -36,25 +34,26 @@
...
 
 	@ManyToMany(mappedBy = "directions")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<UserEntity> users;
+	private Set<UserEntity> users = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/DirectionEntity.java:52]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/InstitutionEntity.java [hunk 5]
```diff
@@ -46,10 +44,18 @@
 	@OneToMany(mappedBy = "mainInstitution")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<UserEntity> mainUsersInstitution;
+	private Set<UserEntity> mainUsersInstitution = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/InstitutionEntity.java:47]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/InstitutionEntity.java [hunk 6]
```diff
@@ -46,10 +44,18 @@
...
 
 	@ManyToMany(mappedBy = "institutions")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<UserEntity> users;
+	private Set<UserEntity> users = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/InstitutionEntity.java:52]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java [hunk 7]
```diff
@@ -73,7 +71,7 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<DirectionEntity> directions;
+	private Set<DirectionEntity> directions = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java:74]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java [hunk 8]
```diff
@@ -83,7 +81,7 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<TagEntity> tags;
+	private Set<TagEntity> tags = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java:84]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java [hunk 9]
```diff
@@ -93,23 +91,27 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<SourceEntity> sources;
+	private Set<SourceEntity> sources = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/PostEntity.java:94]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/PostTypeEntity.java [hunk 10]
```diff
@@ -35,10 +33,11 @@
 	@OneToMany(mappedBy = "type")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<PostEntity> posts;
+	private Set<PostEntity> posts = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/PostTypeEntity.java:36]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/RegionEntity.java [hunk 11]
```diff
@@ -35,10 +33,11 @@
 	@OneToMany(mappedBy = "region")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<CityEntity> cities;
+	private Set<CityEntity> cities = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/RegionEntity.java:36]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/RoleEntity.java [hunk 12]
```diff
@@ -35,8 +36,9 @@
 	@ToString.Exclude
 	private Set<UserEntity> users = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/RoleEntity.java:37]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/SourceEntity.java [hunk 13]
```diff
@@ -37,20 +38,20 @@
...
-
 	@ManyToOne
 	@JoinColumn(name = "user_id")
+	private UserEntity user;
+
+	@ManyToMany(mappedBy = "sources")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private UserEntity user;
+	private Set<PostEntity> posts = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/SourceEntity.java:48]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 14]
```diff
@@ -71,12 +72,12 @@
 	@OneToMany(mappedBy = "author")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<PostEntity> posts;
+	private Set<PostEntity> posts = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:75]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 15]
```diff
@@ -71,12 +72,12 @@
...
 
 	@OneToMany(mappedBy = "author")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<CharityEntity> charities;
+	private Set<CharityEntity> charities = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:80]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 16]
```diff
@@ -86,7 +87,7 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<InstitutionEntity> institutions;
+	private Set<InstitutionEntity> institutions = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:90]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 17]
```diff
@@ -96,7 +97,7 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<RoleEntity> roles;
+	private Set<RoleEntity> roles = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:100]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 18]
```diff
@@ -106,18 +107,22 @@
 	)
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<DirectionEntity> directions;
+	private Set<DirectionEntity> directions = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:110]: do not initialize fields in entity

File: src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java [hunk 19]
```diff
@@ -106,18 +107,22 @@
...
 
 	@OneToMany(mappedBy = "user")
 	@EqualsAndHashCode.Exclude
 	@ToString.Exclude
-	private Set<SourceEntity> sources;
+	private Set<SourceEntity> sources = new HashSet<>();
```
Comment (ponruslan) [inline src/main/java/com/softserveinc/dokazovi/entity/UserEntity.java:115]: do not initialize fields in entity

File: src/test/java/com/softserveinc/dokazovi/mapper/UserMapperTest.java [hunk 20]
```diff
@@ -1,36 +1,81 @@
...
+
+		latestPostEntity = PostEntity.builder()
+				.id(2)
+				.title("Latest Post title")
+				.author(userEntity)
+				.createdAt(latestCreatedAt)
+				.status(PostStatus.PUBLISHED)
+				.build();
+
+		userEntity.getPosts().add(postEntity);
```
Comment (ponruslan) [inline src/test/java/com/softserveinc/dokazovi/mapper/UserMapperTest.java:74]: use userEntity.setPosts(Set.of(postEntity, latestPostEntity)).

**Review by ponruslan — CHANGES_REQUESTED** [review]

**Review by ponruslan — APPROVED** [review]

**Review by melnikovae87 — APPROVED** [review]

**Pull request merged**